code/datums/gamemode/role/changeling.dm

```diff
@@ -0,0 +1,599 @@
+	helptext = "Heals a moderate amount of damage every tick."
+	genomecost = 8
+	verbpath = /obj/item/verbs/changeling/proc/changeling_rapidregen
+
+/datum/power/changeling/armblade
+	name = "Arm Blade"
+	desc = "We transform one of our arms into an organic blade that can cut through flesh and bone."
+	helptext = "The blade can be retracted by using the same verb used to manifest it. It has a chance to deflect projectiles."
+	genomecost = 5
+	verbpath = /obj/item/verbs/changeling/proc/changeling_armblade
+
+
+/datum/role/changeling/proc/EvolutionMenu()
+	set category = "Changeling"
+	set desc = "Level up!"
+
+	if(!usr || !usr.mind || !usr.mind.GetRole(CHANGELING))
+		return
+	src = usr.mind.GetRole(CHANGELING)
+
+	if(!powerinstances.len)
+		for(var/P in powers)
+			powerinstances += new P()
+
+	var/dat = "<html><head><title>Changling Evolution Menu</title></head>"
+
+	//javascript, the part that does most of the work~
+	dat += {"
+
+		<head>
+			<script type='text/javascript'>
+
+				var locked_tabs = new Array();
+
+				function updateSearch(){
+
+
+					var filter_text = document.getElementById('filter');
+					var filter = filter_text.value.toLowerCase();
+
+					if(complete_list != null && complete_list != "")
+						{
+						var mtbl = document.getElementById("maintable_data_archive");
+						mtbl.innerHTML = complete_list;
+					}
+
+					if(filter.value == "")
+						{
+						return;
+					}else{
+
+						var maintable_data = document.getElementById('maintable_data');
+						var ltr = maintable_data.getElementsByTagName("tr");
+						for ( var i = 0; i < ltr.length; ++i )
+						{
+							try{
+								var tr = ltr\[i\];
+								if(tr.getAttribute("id").indexOf("data") != 0)
+									{
+									continue;
+								}
+								var ltd = tr.getElementsByTagName("td");
+								var td = ltd\[0\];
+								var lsearch = td.getElementsByTagName("b");
+								var search = lsearch\[0\];
+								//var inner_span = li.getElementsByTagName("span")\[1\] //Should only ever contain one element.
+								//document.write("<p>"+search.innerText+"<br>"+filter+"<br>"+search.innerText.indexOf(filter))
+								if ( search.innerText.toLowerCase().indexOf(filter) == -1 )
+								{
+									//document.write("a");
+									//ltr.removeChild(tr);
+									td.innerHTML = "";
+									i--;
+								}
+							}catch(err) {   }
+						}
+					}
+
+					var count = 0;
+					var index = -1;
+					var debug = document.getElementById("debug");
+
+					locked_tabs = new Array();
+
+				}
+
+				function expand(id,name,desc,helptext,power,ownsthis){
+
+					clearAll();
+
+					var span = document.getElementById(id);
+
+
+					body = "<table><tr><td>";
+					body +=	"</td><td align='center'>";
+					body +=	"<font size='2'><b>"+desc+"</b></font> <BR>";
+					body +=	"<font size='2'><font color = 'red'><b>"+helptext+"</b></font> <BR>";
+
+					if(!ownsthis)
+					{
+						body += "<a href='?src=\ref[src];P="+power+";mind=\ref[antag];'>Evolve</a>"
```

Comment:
<a href='?src=\ref[src];mind=\ref[antag];(whatever else)'> is how we should approach html links in the future involving role HTML stuff